### PR TITLE
Return better errors on postgres in ApplyMigrations

### PR DESCRIPTION
**Pull request by tomhoule**
A lot of this commit is about taking advantage of
prisma/quaint#301 to gain access to the
underlying database driver in sql-migration-connector.

Rebased and dependent on https://github.com/prisma/prisma-engines/pull/2025

### CODE REVIEW

File: migration-engine/connectors/sql-migration-connector/src/flavour/postgres.rs [hunk 1]
```diff
@@ -114,6 +119,82 @@ impl SqlFlavour for PostgresFlavour {
...
+                            } else {
+                                previous_lines = [
+                                    previous_lines[1],
+                                    previous_lines[2],
+                                    previous_lines[3],
+                                    previous_lines[4],
+                                    line,
+                                ];
+                            }
+                        }
```
Comment (tomhoule) [inline migration-engine/connectors/sql-migration-connector/src/flavour/postgres.rs:175]: This should probably move into its own file.

Comment (pimeys) [reply migration-engine/connectors/sql-migration-connector/src/flavour/postgres.rs:178]: What's going on in here? What should this code be doing?

File: migration-engine/migration-engine-tests/tests/migrations/postgres.rs [hunk 2]
```diff
@@ -214,3 +214,117 @@ fn functions_with_schema_prefix_in_dbgenerated_are_idempotent(api: TestApi) {
...
+        .generated_migration_name
+        .unwrap();
+
+    let err = api
+        .apply_migrations(&migrations_directory)
+        .send_unwrap_err()
+        .to_string()
+        .replace(&migration_name, "<migration-name>");
+
+    let expectation = expect![[r#"
```
Comment (pimeys) [inline migration-engine/migration-engine-tests/tests/migrations/postgres.rs:254]: ach so...

File: migration-engine/migration-engine-tests/tests/migrations/shadow_database_url_configuration.rs [hunk 3]
```diff
@@ -170,24 +171,14 @@ fn shadow_db_not_reachable_error_must_have_the_right_connection_info(api: TestAp
         .to_user_facing();
 
     let assertion = expect![[r#"
-        Error {
```
Comment (pimeys) [inline migration-engine/migration-engine-tests/tests/migrations/shadow_database_url_configuration.rs:173]: 👍🏾

**Review by pimeys — APPROVED** [review]

File: migration-engine/connectors/sql-migration-connector/src/flavour/postgres.rs (hunk 1 above)
Comment (tomhoule) [reply migration-engine/connectors/sql-migration-connector/src/flavour/postgres.rs:178]: It prints the line where the error happened in red, with surrounding context and line numbers.

**Pull request merged**